core/src/main/java/org/apache/calcite/tools/RelBuilder.java

```diff
@@ -495,6 +495,10 @@ public RexNode field(int inputCount, String alias, String fieldName) {
 
   /** Returns a reference to a given field of a record-valued expression. */
   public RexNode field(RexNode e, String name) {
+    if (!e.getType().isStruct()) {
+      throw new IllegalArgumentException("Requested field " + name + " in non-struct expression "
+          + e.toString() + " (type: " + e.getType() + ")");
+    }
```

Comment:
I am wondering where is it better to add this check here or inside `RexBuilder#makeFieldAccess`. Clients who call `RexBuilder` direclty will still get the NPE.

Reply:
I have the impression that the `RelBuilder` build rel trees based on all kinds of assumptions, it assumes that the invoker knows how to behave with the valid operands. We may need to check all the interfaces in `RelBuilder`, this PR is a good start.

Reply:
Thanks for the feedback. Looking again at this code in particular, I have the impression that a simpler fix for this particular issue would be leaving `RelBuilder` untouched and modify this line in `RelDataTypeImpl.getField`:
```
  @Override public RelDataTypeField getField(String fieldName, boolean caseSensitive,
      boolean elideRecord) {
    for (RelDataTypeField field : fieldList) {
// =>
  @Override public RelDataTypeField getField(String fieldName, boolean caseSensitive,
      boolean elideRecord) {
    for (RelDataTypeField field : getFieldList()) {  // change here
```
With this change, the scenario in the Jira would fail with an AssertionError (thrown by `getFieldList()`) instead of NPE.

Reply:
I believe this check should be moved to `org.apache.calcite.rel.type.RelDataTypeImpl#getField`
Then all uses of `RelDataTypeImpl#getField` would get the proper exception rather than NPE.

Reply:
@vlsi we had the same thought :)

Reply:
>With this change, the scenario in the Jira would fail with an AssertionError (thrown by getFieldList()) instead of NPE.

Please use the tailored null check and the proper exception so the user could know which field is wrong. Then user could relate the problematic field name to the input SQL.

Reply:
After a second thought, apart from changing RelDataTypeImpl to have an AssertionError instead of NPE, I added a specific check in RexBuilder, in order to have the same behavior in both public methods `makeFieldAccess`

Reply:
An assertion in `RelDataTypeImpl` is enough. What is the purpose of duplicating the same check in multiple places?

Reply:
Just in case, there's `org.apache.calcite.piglet.DynamicTupleRecordType` which supports `getField(String)`, and it is not struct at the same time.
That is why I'm inlined `isStruct()` check must not be added.

Reply:
The reason would be having the same "tailored check" in both RexBuilder methods:
```
public RexNode makeFieldAccess(RexNode expr, String fieldName, boolean caseSensitive) {...}
public RexNode makeFieldAccess(RexNode expr,int i) {...}
```
Also, there is the fact that assertions can be deactivated, so I would lean towards an `IllegalArgumentException` in this case.
Finally, we do know that `RelDataTypeImpl` has some assertions in place, but it is not defined the behavior that other `RelDataType` implementions may have in this situation.

Having said that, if the community thinks these checks are redundant, I am not against removing them, keeping both `RexBuilder#makeFieldAccess` methods as they are; and in case of incorrect calls they will both eventually fail with `AssertionError` thrown by `RelDataTypeImpl` when the try to call `getField` or `getFieldList`

Reply:
@vlsi if I am not mistaken, `org.apache.calcite.piglet.DynamicTupleRecordType` inherits `isStruct` from `org.apache.calcite.rel.type.DynamicRecordTypeImpl`:
```
  @Override public boolean isStruct() {
    return true;
  }
```
But, what you say is a valid point, if we have the assumption that a non-struct type can support getField, then `isStruct()` check must not be added. I'm not sure if this scenario makes sense at all, but just in case perhaps the most conservative approach would be removing the  `isStruct()`  check and just rely on the `AssertionError` thrown by `RelDataTypeImpl`.
Honestly, this is a minor change, I am open to any suggestion.

Reply:
I'm inclined it should be `IllegalStateException` since any argument to `RelDataTypeImpl#getField` would result in an exception.

Reply:
I proposed `IllegalArgumentException`, considering the first argument to be the illegal one (`RexNode expr`, having a non-struct type). But I can see `IllegalStateException` fitting here as well.

@zabetak @danny0405 what are your thoughts regarding the new proposal? Are you against adding the check in `RexBuilder#makefieldAccess` (hence having an `AssertionError` from `RelDataTypleImpl`); or for adding check (in which case, would you see it as an `IllegalArgumentException` or `IllegalStateException`)?

Reply:
Basically, I do not like to make `RexBuilder` aware of the ways `getField` can fail.
In other words, it is not RexBuilder's business to tell if `non-sturcts` are allowed to have fields or not.

I've had a very similar discussion with @julianhyde in https://issues.apache.org/jira/browse/CALCITE-4217, and the takeaway was "don't try to make sense of struct vs fieldList".

That is why I believe the necessary and sufficient solution here is:
1) Delegate the proper exception message to `org.apache.calcite.rel.type.RelDataType#getField(String fieldName, ...)` implementations. In other words, if someone implements `RelDataType`, then it is their business to throw the appropriate exception.

2) `org.apache.calcite.rex.RexBuilder#makeFieldAccess(org.apache.calcite.rex.RexNode, int)` is slightly different since it can't delegate to `RelDataType`, and it has to call `type.getFieldList()`. It could be left as is, and the type implementation would throw "type ... has no fields" (which might be enough).
If you want to add clarification with the problematic `RexNode`, then you could use

```java
try {
  final List<RelDataTypeField> fields = type.getFieldList();
} catch (RuntimeException e) {
  e.addSuppressed(new Throwable("Unable to get field " + i + " from rexNode " + rexNode");
  throw e
}
```

The important point is in both cases `RexBuilder` makes no assumptions on `struct` vs `non-struct`.

Reply:
Ah, indeed, `RelCrossType` is a case when `isStruct == false` and `getFieldList()` returns non-null.

Reply:
> Ah, indeed, RelCrossType is a case when isStruct == false and getFieldList() returns non-null.
Well spotted, thanks for the feedback. I'll revert the changes in RexBuilder

Reply:
Committed new proposal taking into account @vlsi 's input